### PR TITLE
Bug 1884632: Adding BYOK disk encryption through DES

### PR DESCRIPTION
**Pull request by kwoodson**
This pull requests adds fields to the v1beta1 types to suppert disk encryption sets.

We have customers requesting that we provide OS disk encryption.  AKS is using DES for OS level as well as PVC support.  The following PR was an attempt to add the necessary fields for bring-your-own-key encryption using DiskEncryptionSets.

This is high priority for the ARO team and we would like to get the work going.

### CODE REVIEW

**Comment by kwoodson** [comment]
@ingvagabund, @enxebre 
I attempted to add a few required fields to the `ManagedDisk`.  I was able to build this and place this image into my cluster but when I went to create a machine on my 4.4.10 cluster I had the following errors:
```
oc logs machine-api-controllers-77b77ccc8f-vnmwj  -c machine-controller
...
I0813 20:56:18.403374       1 controller.go:169] kwtest-w6n4r-worker-eastus1-des-9srq8: reconciling Machine
I0813 20:56:18.403412       1 actuator.go:201] kwtest-w6n4r-worker-eastus1-des-9srq8: actuator checking if machine exists
E0813 20:56:18.403952       1 controller.go:272] kwtest-w6n4r-worker-eastus1-des-9srq8: failed to check if machine exists: failed to create scope: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal string into Go struct field ManagedDiskParameters.osDisk.managedDisk.diskEncryptionSet of type v1beta1.DiskEncryptionSetParameters%!(EXTRA string=failed to get machine config)
E0813 20:56:18.404064       1 controller.go:258] controller-runtime/controller "msg"="Reconciler error" "error"="failed to create scope: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal string into Go struct field ManagedDiskParameters.osDisk.managedDisk.diskEncryptionSet of type v1beta1.DiskEncryptionSetParameters%!(EXTRA string=failed to get machine config)"  "controller"="machine_controller" "request"={"Namespace":"openshift-machine-api","Name":"kwtest-w6n4r-worker-eastus1-des-9srq8"}
```
I must have missed something.  Do you have any suggestions?

**Comment by kwoodson** [comment]
@enxebre @ingvagabund PTAL

**Comment by kwoodson** [comment]
@ingvagabund When running `make generate` the `zz_generated.deepcopy.go` was not generated.  This might be due to my local setup.  I hand patched it for the time being but it should be generated.

**Comment by kwoodson** [comment]
I had to grant `Reader` role on the DES for the service principal and then with this latest update I was able to get this to work:
```
oc get machines
...
kwtest-w6n4r-worker-eastus1-des-9xqdr   Running   Standard_D2s_v3   eastus   1      6m4s
```

Command line:
```
az vm show -n kwtest-w6n4r-worker-eastus1-des-9xqdr -g aro-w5o8qzfu
...
    "osDisk": {
      "caching": "ReadWrite",
      "createOption": "FromImage",
      "diffDiskSettings": null,
      "diskSizeGb": 128,
      "encryptionSettings": null,
      "image": null,
      "managedDisk": {
        "diskEncryptionSet": {
          "id": "/subscriptions/225e02bc-43d0-43d1-a01a-17e584a4ef69/resourceGroups/disk-encryption/providers/Microsoft.Compute/diskEncryptionSets/aro-de",
          "resourceGroup": "disk-encryption"
        },
        "id": "/subscriptions/225e02bc-43d0-43d1-a01a-17e584a4ef69/resourceGroups/ARO-W5O8QZFU/providers/Microsoft.Compute/disks/kwtest-w6n4r-worker-eastus1-des-9xqdr_OSDisk",
```

From the portal I was able to confirm the encryption:
```
kwtest-w6n4r-worker-eastus1-des-9xqdr_OSDisk
128 GiB
Premium SSD
SSE with CMK               <----- SSE (storage service encryption) with CMK (customer managed key)
Read/write
```

**Comment by michaelgugino** [comment]
Make generate scripts were broken.  We'll need to re-run the make generate on this patch set after this merges: https://github.com/openshift/cluster-api-provider-azure/pull/159

Otherwise, this change LGTM

**Comment by ingvagabund** [comment]
> I had to grant Reader role on the DES for the service principal 

Worth documenting the additional role

File: pkg/apis/azureprovider/v1beta1/types.go
Comment (abhinavdahiya) [inline pkg/apis/azureprovider/v1beta1/types.go:411]: it would be nice if we can provide details on what kind of permissions are required to use this disk encryption set for this machine.

File: pkg/apis/azureprovider/v1beta1/types.go
Comment (abhinavdahiya) [inline pkg/apis/azureprovider/v1beta1/types.go:416]: Would be nice to include some documentation on the the expected form of the ID, Using Azure resource IDs implies that this resource can be from any subscription, any resource group.

Since this is for supporting customer managed encryption, it is fair to say we need to support this resource from resource group other than the cluster's but what about the subscription.

File: pkg/apis/azureprovider/v1beta1/types.go
Comment (kwoodson) [reply pkg/apis/azureprovider/v1beta1/types.go:411]: The service principal that is used with the cluster will require the reader role on the disk encryption set.

File: pkg/apis/azureprovider/v1beta1/types.go
Comment (kwoodson) [reply pkg/apis/azureprovider/v1beta1/types.go:416]: Here is an example of an a disk encryption set ID.  Azure resource IDs include a path-like system so that it is clear where the object is located, what the object is, and its name.
```
/subscriptions/<subscriptionid>/resourceGroups/disk-encryption/providers/Microsoft.Compute/diskEncryptionSets/aro-de
```
According to the documentation here https://docs.microsoft.com/en-us/azure/virtual-machines/linux/disk-encryption-key-vault#create-a-key-vault it states:
```
Your key vault and VMs must be in the same subscription. Also, to ensure that encryption secrets don't cross regional boundaries, Azure Disk Encryption requires the Key Vault and the VMs to be co-located in the same region. Create and use a Key Vault that is in the same subscription and region as the VMs to be encrypted.
```

**Comment by kwoodson** [comment]
Rebased with #158, `make generate`, and pushed.  Let's go CI bot!

**Comment by kwoodson** [comment]
I guess I need to add the `UPSTREAM: <carry>:` git commit message.

**Comment by kwoodson** [comment]
Looks like flakes under both e2e-azure and the operator:
```
11/42
- ingress
- dns
- image-registry
- monitoring
etc
```
```
3/2465
Failing tests:

[sig-cli] oc adm must-gather runs successfully [Suite:openshift/conformance/parallel]
[sig-cli] oc adm must-gather runs successfully with options [Suite:openshift/conformance/parallel]
```

**Comment by kwoodson** [comment]
@elmiko @michaelgugino  Any ideas why this continues to fail?  It looks like must-gather is failing.  This PR only adds fields but isn't actually exercised.  Is this inherently flaky?

What do we do to get this to pass?

File: pkg/apis/azureprovider/v1beta1/types.go
Comment (enxebre) [inline pkg/apis/azureprovider/v1beta1/types.go:412]: Is this ID used at all? I can't see where.

Comment (kwoodson) [reply pkg/apis/azureprovider/v1beta1/types.go:412]: I believe it is used here:
https://github.com/openshift/cluster-api-provider-azure/pull/158/files#diff-79058a3142fd624b3690ff1dc72c0d2cR193

Comment (kwoodson) [reply pkg/apis/azureprovider/v1beta1/types.go:412]: @enxebre, sorry I thought this referenced the `DiskEncryptionSet.ID`.  I included this `ID` because it is found inside of the `ManagedDiskParameters` here https://github.com/Azure/azure-sdk-for-go/blob/master/services/compute/mgmt/2019-12-01/compute/models.go#L6507.  In case it is ever parsed or assigned the struct here would have a reference to it.

Comment (michaelgugino) [reply pkg/apis/azureprovider/v1beta1/types.go:412]: Let's not leave in unused fields so we don't have to worry about a user putting something wrong in there.

Comment (kwoodson) [reply pkg/apis/azureprovider/v1beta1/types.go]: Removed

**Comment by kwoodson** [comment]
woohoo, finally passed 1 of the tests.   Now let's keep trying.

/test e2e-azure-operator

**Comment by enxebre** [comment]
@kwoodson is there any e2e testing PR planned for this? has this manually been verified to work?

**Comment by kwoodson** [comment]
@enxebre 

- I have disk encryption with byok setup and working in a cluster right now.  
- I don't have any specific e2e testing for this as it requires additional azure objects to be created and used within the azure subscription (disk encryption set, azure vault, encryption key, reader role assignment for des).  
- This change does not require any permission changes unless the disk encryption feature is used.  The error found in the machine api controller is the following:
```
"The client '<uuid>' with object id '<uuid>' has permission to perform action 'Microsoft.Compute/virtualMachines/write' on scope '/subscriptions/<subid>/resourceGroups/aro-y72yne1j/providers/Microsoft.Compute/virtualMachines/kwdemo-demo-worker-eastus1-des-w4v9n'; however, it does not have permission to perform action 'read' on the linked scope(s) '/subscriptions/<subid>/resourceGroups/disk-encryption/providers/Microsoft.Compute/diskEncryptionSets/aro-de' or the linked scope(s) are invalid."
```

Please let me know if you have any other questions.  I would like to get this merged when you see fit.  This PR is pretty straightforward and only adds a couple of fields to existing objects.  Then passes them along if they are set.

Thanks

File: pkg/apis/azureprovider/v1beta1/types.go
Comment (jim-minter) [inline pkg/apis/azureprovider/v1beta1/types.go:405]: What's the benefit of renaming ManagedDisk -> ManagedDiskParameters?  Clearly it doesn't affect the serialisation but it'll cause a rename for any Go developer who initialises a ManagedDisk{} when they revendor (e.g. see the stubs.go change below).  I'm only seeing downside.

File: pkg/apis/azureprovider/v1beta1/types.go
Comment (jim-minter) [inline pkg/apis/azureprovider/v1beta1/types.go:415]: I'd consider resourceID rather than id - may be a little clearer.

Comment (jim-minter) [reply pkg/apis/azureprovider/v1beta1/types.go:415]: also note elsewhere this API already uses resourceID...

File: pkg/apis/azureprovider/v1beta1/types.go
Comment (kwoodson) [reply pkg/apis/azureprovider/v1beta1/types.go:405]: I was keeping inline with what is in upstream. 
OSDisk: https://github.com/Azure/azure-sdk-for-go/blob/master/services/compute/mgmt/2019-12-01/compute/models.go#L6700
ManagedDiskParameters: https://github.com/Azure/azure-sdk-for-go/blob/master/services/compute/mgmt/2019-12-01/compute/models.go#L6507

If we prefer to keep our own naming then I can revert.

File: pkg/apis/azureprovider/v1beta1/types.go
Comment (kwoodson) [reply pkg/apis/azureprovider/v1beta1/types.go:415]: Again, aligning with what's in the upstream repository:
https://github.com/Azure/azure-sdk-for-go/blob/master/services/compute/mgmt/2019-12-01/compute/models.go#L2408

**Comment by enxebre** [comment]
>This change does not require any permission changes unless the disk encryption feature is used. The error found in the machine api controller without the required permission looks like the following:

So you are saying the field this PR is adding is unusable without manually changing permissions?
If so this https://github.com/openshift/machine-api-operator/blob/master/install/0000_30_machine-api-operator_00_credentials-request.yaml#L56-L73 needs to be updated to give enough permissions to consume anything we support in the API

**Comment by kwoodson** [comment]
@enxebre 

> So you are saying the field this PR is adding is unusable without manually changing permissions?
> If so this https://github.com/openshift/machine-api-operator/blob/master/install/0000_30_machine-api-operator_00_credentials-request.yaml#L56-L73 needs to be updated to give enough permissions to consume anything we support in the API

This field is basically a no-op unless using more features inside of Azure.  This field does not do anything unless you create a disk encryption set object, a key vault object, import a key into the key vault, and then grant the cluster's service principal read access to the disk encryption set.  Once these steps have been completed, the resourceID can be passed to the machine/machineset to use this feature.

Since these objects are currently not created or managed by Openshift, I do not believe the credentials should be managed by the openshift-machine-api.  If there are plans down the road to implement an operator to manage the life cycle of those objects then I would suggest the permissions be managed by that operator.

Let's catch up on slack if that helps in the conversation.  I have reached out internally but understand that our time zones overlap in my morning.

**Comment by abhinavdahiya** [comment]
> This field is basically a no-op unless using more features inside of Azure. This field does not do anything unless you create a disk encryption set object, a key vault object, import a key into the key vault, and then grant the cluster's service principal read access to the disk encryption set.

This is making the assumption that the cluster service principal is the being passed to all the cluster operators like on ARO when a lot of openshift clsuters have per operator service principals created by the cluster. In which case users will have to update the the machine-api operator's service principal to use this new field.

**Comment by kwoodson** [comment]
@abhinavdahiya 

> This is making the assumption that the cluster service principal is the being passed to all the cluster operators like on ARO when a lot of openshift clsuters have per operator service principals created by the cluster. In which case users will have to update the the machine-api operator's service principal to use this new field.

I understand that there is a difference between OCP on Azure and ARO in terms of authorization.  Since the openshift-machine-api is not managing the disk encryption sets, the key vaults, and for security reasons customers might not want the openshift-machine-api mutating the access permissions to their encryption keys.

Could you or @enxebre please explain to me what permission you would like me to add and how it will work with the extra infrastructure required?

**Comment by michaelgugino** [comment]
> > This field is basically a no-op unless using more features inside of Azure. This field does not do anything unless you create a disk encryption set object, a key vault object, import a key into the key vault, and then grant the cluster's service principal read access to the disk encryption set.
> 
> This is making the assumption that the cluster service principal is the being passed to all the cluster operators like on ARO when a lot of openshift clsuters have per operator service principals created by the cluster. In which case users will have to update the the machine-api operator's service principal to use this new field.

Yes, this is the case, except in Azure, we're just giving everyone the contributor role by default: https://github.com/openshift/machine-api-operator/blob/master/install/0000_30_machine-api-operator_00_credentials-request.yaml#L73

So, in this case, there are no additional permissions we need to add because the machine-api has them all already.  We're hoping to scope this down for 4.7.

**Comment by abhinavdahiya** [comment]
> @abhinavdahiya
> 
> > This is making the assumption that the cluster service principal is the being passed to all the cluster operators like on ARO when a lot of openshift clsuters have per operator service principals created by the cluster. In which case users will have to update the the machine-api operator's service principal to use this new field.
> 
> I understand that there is a difference between OCP on Azure and ARO in terms of authorization. Since the openshift-machine-api is not managing the disk encryption sets, the key vaults, and for security reasons customers might not want the openshift-machine-api mutating the access permissions to their encryption keys.
> 
> Could you or @enxebre please explain to me what permission you would like me to add and how it will work with the extra infrastructure required?

The machine-api only needs read permissions on the disk encryption sets that are being specified, so I don't think we are asking for some way to define how to do that for people in OCP and ARO, or restrict use of this.

**Comment by abhinavdahiya** [comment]
> > > This field is basically a no-op unless using more features inside of Azure. This field does not do anything unless you create a disk encryption set object, a key vault object, import a key into the key vault, and then grant the cluster's service principal read access to the disk encryption set.
> > 
> > 
> > This is making the assumption that the cluster service principal is the being passed to all the cluster operators like on ARO when a lot of openshift clsuters have per operator service principals created by the cluster. In which case users will have to update the the machine-api operator's service principal to use this new field.
> 
> Yes, this is the case, except in Azure, we're just giving everyone the contributor role by default: https://github.com/openshift/machine-api-operator/blob/master/install/0000_30_machine-api-operator_00_credentials-request.yaml#L73
> 
> So, in this case, there are no additional permissions we need to add because the machine-api has them all already. We're hoping to scope this down for 4.7.

The contributor role is only on the resource group where the cluster was installed. So the only disk encryption sets you are use are in the same resource group as the cluster. but the API allows ANY.

most people have other teams managing these disk encryption sets, and these are never going to be in the same resource group as the cluster. And also since we create a resource group for our clusters to assert management boundary there would be no way to get these machines encrypted with disk-encryption-sets from other resource group during installation.

**Comment by michaelgugino** [comment]
> most people have other teams managing these disk encryption sets, and these are never going to be in the same resource group as the cluster. And also since we create a resource group for our clusters to assert management boundary there would be no way to get these machines encrypted with disk-encryption-sets from other resource group during installation.

This may all be true, but there's no mechanism today to allow any custom permissions on any cloud provider.  The user would have to grant those extra permissions out of band.  So as far as what I would consider 'normally supported' credentials, there's nothing blocking this patch set.  The cloud-credential-operator is not resource group aware.

File: pkg/apis/azureprovider/v1beta1/types.go
Comment (michaelgugino) [reply pkg/apis/azureprovider/v1beta1/types.go:415]: Yeah, I like aligning with upstream.

Comment (michaelgugino) [reply pkg/apis/azureprovider/v1beta1/types.go:415]: +1 to leave as-is

File: pkg/apis/azureprovider/v1beta1/types.go
Comment (michaelgugino) [reply pkg/apis/azureprovider/v1beta1/types.go:405]: We're not renaming the field name, just the type name, which is not exposed to users and won't require any changes.  +1 to aligning with upstream.

**Review by michaelgugino — APPROVED** [review]
/lgtm

**Comment by enxebre** [comment]
@kwoodson can we please include a comment in the git description and in the new API field clarifying the expectations.
E.g: This assumes the disk encryption set leaves in the same resource group than the VM where the Cloud Credentials Operator granted permissions. Otherwise additional reader role permissions needs to be granted on the Service principal.

Also given the lack of automated testing for this feature it'd be nice to follow the "Bug Fix PR Pre-Merge confirmation testing" approach for this PR https://issues.redhat.com/browse/DPTP-660. Not a blocker to me though.

**Comment by kwoodson** [comment]
@enxebre @michaelgugino  
I updated the PR's commit message to make a note about the access as @enxebre requested.

**Review by michaelgugino — APPROVED** [review]
/lgtm

**Comment by enxebre** [comment]
/approve
/hold
for main branch to open for 4.7.

**Comment by kwoodson** [comment]
I've asked @mgahagan73 to run through a setup and test of this for verification.  He is currently working his way through and will add a note when he finishes.

**Comment by mgahagan73** [comment]
I have verified this feature by creating a disk encryption set and keyvault, and then creating another machineset which used the specified disk encryption set. I was then able to confirm that the new vm created by the machineset  is using the customer provided key which I created.  QE can test this feature going forward based on the instructions provided by @kwoodson

**Comment by kwoodson** [comment]
```
1/2475 Tests Failed.

Failing tests:

[sig-storage] In-tree Volumes [Driver: gluster] [Testpattern: Dynamic PV (ntfs)][sig-windows] subPath should support readOnly directory specified in the volumeMount [Skipped:ibmcloud] [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir-link] [Testpattern: Inline-volume (default fs)] subPath should support file as subpath [LinuxOnly] [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] In-tree Volumes [Driver: nfs] [Testpattern: Dynamic PV (ext4)] volumes should allow exec of files on the volume [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] In-tree Volumes [Driver: vsphere] [Testpattern: Inline-volume (default fs)] volumes should allow exec of files on the volume [Suite:openshift/conformance/parallel] [Suite:k8s]

Writing JUnit report to /tmp/artifacts/junit/junit_e2e_20200923-222859.xml

error: 8 fail, 994 pass, 1417 skip (28m52s)
error: failed to execute wrapped command: exit status 1
```

**Comment by kwoodson** [comment]
@elmiko @enxebre  @michaelgugino Seems as though this test is all over the place.  It went from 1 failure to 24 to 81 on unrelated topics to this PR.  Is this normally the process?
```
81/2474 Tests Failed
```

**Comment by elmiko** [comment]
> @elmiko @enxebre @michaelgugino Seems as though this test is all over the place. It went from 1 failure to 24 to 81 on unrelated topics to this PR. Is this normally the process?

i have seen similar types of failures, although i don't recognize the issue in the current failure you are seeing. we have also seen a bunch of strange behavior during the deployment phase of some tests recently. i'm not sure what the best course of action here is, but it might be worth asking on #4-dev-triage to see if others know about this.

**Comment by michaelgugino** [comment]
@kwoodson that azure test is not required.  What you need is a BZ targeted at 4.6.

**Comment by kwoodson** [comment]
@michaelgugino Thanks, I'll ping in the AM.

**Comment by michaelgugino** [comment]
/test e2e-azure

Testing CI.

**Comment by JoelSpeed** [comment]
@kwoodson If this merges into 4.6, you will need to revendor this repository into the MAO repo so that the updated types are included in the webhooks (else the defaulting will drop any fields it is unaware of)

**Comment by JoelSpeed** [comment]
/bugzilla refresh
/hold cancel

Once this is merged, we need to add a PR to revendor this repo into MAO before the BZ will be able to be verified

**Comment by mjudeikis** [comment]
@JoelSpeed @kwoodson with the amount of azure-operator fails where, are we sure this is a flake? O_O I'm but worried this is is a flake - we have a bigger problems here.

**Comment by JoelSpeed** [comment]
@mjudeikis I've had a look through the azure failures for this PR and sadly, there's not very much consistency between the failures and unfortunately, this is what we have come to expect from Azure CI lately. Pretty much every PR to this repo will retest for 4/5 days before it eventually passes.

Azure seems to be having issues providing VMs at the moment. I even saw this yesterday while using a dev cluster, the VM just never left the creating state and then after 90 minutes Azure gave up on it.

**Comment by kwoodson** [comment]
@mjudeikis This test is completely broken.  I've seen it pass a single time but each failure is inconsistent like @JoelSpeed has mentioned.  I've seen quite a few failures with `must-gather` tests, nodes not provisioning, `gluster`, etc.  I'm currently looking at the failures in the background to see if we can't fix some of them.

**Comment by kwoodson** [comment]
Looks like this failed on the `Running on Spot` test.  
```
 Timed out after 1800.000s.
[4]   Unexpected error:
[4]       <*errors.errorString | 0xc0002faef0>: {
[4]           s: "\"ci-op-zkn51jdw-be5ef-6fbrrrqwbv\": not all Machines are running: 2 of 3",
[4]       }
[4]       "ci-op-zkn51jdw-be5ef-6fbrrrqwbv": not all Machines are running: 2 of 3
[4]   occurred
```

**Comment by JoelSpeed** [comment]
> Looks like this failed on the `Running on Spot` test.
> 
> ```
>  Timed out after 1800.000s.
> [4]   Unexpected error:
> [4]       <*errors.errorString | 0xc0002faef0>: {
> [4]           s: "\"ci-op-zkn51jdw-be5ef-6fbrrrqwbv\": not all Machines are running: 2 of 3",
> [4]       }
> [4]       "ci-op-zkn51jdw-be5ef-6fbrrrqwbv": not all Machines are running: 2 of 3
> [4]   occurred
> ```

I've seen this one before (many times), it means that the Machines didn't all start up. If my analysis of the situation is correct here, this is because we create a new MachineSet at the beginning of the test and wait for all of the Machines to be `Running`, except quite often we don't see them all come us a running because it takes a while and the test times out before the final machine has booted and started. I have been assuming this was Azure just being slow to provide machines?

**Comment by kwoodson** [comment]
@JoelSpeed I think specifically the spot instances are less reliable.  I have run this test inside of our team's subscription yesterday and the test suite passed 3 times in a row.  The issue could be in the PR's code which I don't believe it is due to all other tests passing but due to the reliability of spot instances.

**Comment by JoelSpeed** [comment]
> I have run this test inside of our team's subscription yesterday and the test suite passed 3 times in a row.

The entire suite or the spot test in particular? 

> reliability of spot instances

So you're thinking that on Azure spot instances are less reliable and as such are being taken away before they become ready? Or just spot instances take longer to start up? I can try and do some research into this. I'll also try to take a look at the test code before tomorrow's meeting in case there's any obvious changes that can be made to this test to improve reliability

**Comment by kwoodson** [comment]
> The entire suite or the spot test in particular?
I have seen spot instances take longer to come up or not come up at all.

I'm going to request access to the CI environment in order to test there and attempt to reproduce.

**Comment by mjudeikis** [comment]
Upstream upstream: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/980 
Upstream api issue: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/982

**Pull request merged**
